### PR TITLE
EVG-18444: Allow nested folders in GraphQL tests

### PR DESCRIPTION
**Pull request by minnakt**
EVG-18444

#### Context
It has become increasingly difficult for the UI team to manage the GraphQL tests. Many of the tests share a single `data.json` file, which are often quite long and hard to parse. Additionally, when these data files are modified to accommodate for new tests, it's not uncommon for other tests to break. 

This PR proposes a new structure for the GraphQL tests, which can be outlined as the following:
```
.
├── ...
├── resolver/               # Folder representing a resolver, e.g. query
│   ├── field/              # Folder representing a field on a resolver, e.g. mainlineCommits
│   ├──── queries/          # Folder containing the query files (.graphql)
│   ├──── data.json         # Data for the tests in this directory
│   └──── results.json      # Results for the tests in this directory
└── ...
```
This format will allow us to organize our tests in a more GraphQL-y way. It will also allow us to share data files between a fewer number of tests, which should help to prevent them from growing large and unmanageable.

#### Changes in this PR
In this PR, I only organized the `version` and `task` tests to use this new format so that the changes are easier to review. The test runner has been modified to support both the new and old format for the time being (see changes in [service/graphql/graphql_integration_test.go](https://github.com/evergreen-ci/evergreen/compare/main...minnakt:EVG-18444?expand=1#diff-ae03c44435c28b39739f05f35fb93d7436ec9aacadee1b336083978a49b70bcb)). 

### Testing
GraphQL tests should pass

#### Does this need documentation?
No

### CODE REVIEW

File: service/graphql/graphql_integration_test.go [hunk 1]
```diff
@@ -35,14 +35,36 @@ func TestAtomicGQLQueries(t *testing.T) {
...
+					Settings:    settings,
+					ServerURL:   server.URL,
+				}
+				t.Run(state.Directory, graphql.MakeTestsInDirectory(&state, pathToTests))
+			} else {
+				runOuterDirectoryTests = true
+			}
+		}
+
+		// The following code can be deleted once all tests have been moved to the new folder structure.
```
Comment (malikchaya2) [inline service/graphql/graphql_integration_test.go:57]: Can we please create a ticket for the follow up and put the ticket number in the comment?

**Review by malikchaya2 — APPROVED** [review]
approved with nit

**Comment by minnakt** [comment]
> I was wondering what you thought about adding flexibility where specifying an additional data.json is optional and would override the shared data.json as needed. This might be easier to use when the same seed data doesn't cause issues between tests.

That's an interesting idea! It would have limited application since most of the tests would fall under the `query` or `mutation` parent folders and  there wouldn't be a reason to share data between them, but it could work for other folders. I think we can keep this is mind for a future ticket once all of the tests have been organized because it would probably be easier to modify the test runner when all of the tests follow the same format.

File: service/graphql/graphql_integration_test.go [hunk 2]
```diff
@@ -35,14 +35,36 @@ func TestAtomicGQLQueries(t *testing.T) {
...
+				state := graphql.AtomicGraphQLState{
+					TaskLogDB:   model.TaskLogDB,
+					TaskLogColl: model.TaskLogCollection,
+					Directory:   filepath.Join(dir.Name(), entry.Name()),
+					Settings:    settings,
+					ServerURL:   server.URL,
+				}
+				t.Run(state.Directory, graphql.MakeTestsInDirectory(&state, pathToTests))
+			} else {
+				runOuterDirectoryTests = true
```
[the remark this answers is not in the document]
Comment (minnakt) [reply service/graphql/graphql_integration_test.go]: Oops sorry, I messed up 🥲 I updated these conditionals

File: service/graphql/graphql_integration_test.go [hunk 3]
```diff
@@ -35,14 +35,36 @@ func TestAtomicGQLQueries(t *testing.T) {
...
+			}
+		}
+
+		// The following code can be deleted once all tests have been moved to the new folder structure.
+		if runOuterDirectoryTests {
```
[the remark this answers is not in the document]
Comment (minnakt) [reply service/graphql/graphql_integration_test.go]: Sure, I put some more folders into the `patch` directory (they don't belong there but I can revert it), and you can either run them locally or check the results on the patch

**Comment by minnakt** [comment]
> Do you know why we see an increase in test results with the new format? I'm able to see that by looking at the test result table in Spruce.

Answered in DMs but to answer for public visibility: we record the Pass/Fail status of each entire directory, so because there are more directories, there are also more test results

**Review by khelif96 — COMMENTED** [review]
What if we took this one further and organized our queries by queries and mutations. 

One parent folder for all queries
and one parent folder for all mutations

File: service/graphql/graphql_integration_test.go [hunk 4]
```diff
@@ -35,14 +35,34 @@ func TestAtomicGQLQueries(t *testing.T) {
...
+					state = graphql.AtomicGraphQLState{
+						TaskLogDB:   model.TaskLogDB,
+						TaskLogColl: model.TaskLogCollection,
+						Directory:   filepath.Join(dir.Name(), entry.Name()),
+						Settings:    settings,
+						ServerURL:   server.URL,
+					}
+				}
+				t.Run(state.Directory, graphql.MakeTestsInDirectory(&state, pathToTests))
+			}
```
[the remark this answers is not in the document]
Comment (minnakt) [reply service/graphql/graphql_integration_test.go:64]: I think it's actually okay for there to be stranded files; for example `results.json` and `data.json`. Sometimes you also get hit annoyingly by the `.DS_Store` :\

**Comment by minnakt** [comment]
> What if we took this one further and organized our queries by queries and mutations.
> 
> One parent folder for all queries and one parent folder for all mutations

yup this is the plan! I was going to make some tickets for doing incrementally over time. The query and mutation ones should be pretty easy though, I could try to do them next sprint! :0
